### ISSUE TITLE
KrigingSolver is indexing pdomain instead pdata for searching

### ISSUE THREAD

**Issue by exepulveda** [JuliaEarth/KrigingEstimators.jl#17]
In the preprocess function, pdomain is used to index the neighbours, which conceptualy is not correct.
kriging should index pdata and search k neighbours from pdata instead with no need of masking previously estimated locations.

Currently, it is not clear to me but it seems with datamap Kriging solver is trying to map locations in pdata to pdomain and mask those mapped locations as estimated. This may give different results because pdomain locations may not necessarily match to pdata mapped location  and the covariances may not be the right.

I will propose a fix for this.

**Comment by juliohm** [comment]
I agree we should be indexing the `pdata` instead of the `pdomain` :+1: Based on that discussion we had in the other issue, simple Kriging coincides with the conditional mean of the process at the point, but other variants of Kriging are not equivalent to this conditional mean. So adding the previously estimated points to the set can compromise the estimates.

I also like the idea of eliminating this `mask` trick at some point in the future. Given that now we will be constructing the kdtree directly on pdata, which is fixed and won't be updating the set of estimated points, I think we can avoid this mask issue :100: 

You are correct that the `datamap` is constructed prior to the solution of the problem. It is just a dict mapping the locations of the domain and the data. It basically assigns the initial observations to the grid. There are a few mappers available, I don't know if I followed your comment about the covariances. Could you please elaborate?

**Comment by exepulveda** [comment]
If you move a sample to the grid, you loose the actual sample location. The covariance is function of the distance, and as the distance may change, the covariance may change as well. In coarse grids, the difference can be significant. Also, what would happen if there are two or more samples mapped at the same grid cell?

**Comment by juliohm** [comment]
All valid points. We tried to write different mapping strategies to accommodate some of these issues, but we certainly need to rethink if there is a better way to avoid these direct assignments. This is an issue that will require some brainstorming of the EstimationProblem API which is even more general than the Kriging case.

**Comment by exepulveda** [comment]
And it is urgent if we want this package to be a real alternative to well proven ones. We need to ensure that Kriging works as it should be.

**Comment by juliohm** [comment]
Yes, there is a lot of work to make the project fulfill the requirements of
mining applications. I've been quite busy trying to build the foundations
and data structures of the project, but couldn't find much time to fix
issues regarding specific solvers. It is nice to have you here to help.

Em sex., 31 de jul. de 2020 às 08:30, exepulveda <notifications@github.com>
escreveu:

**Comment by exepulveda** [comment]
And here I am to help 🎉
